Atmosphere LUT parameterization improvements

mate-h opened the pull request:
# Objective

- Fix the atmosphere LUT parameterization in the aerial -view and sky-view LUTs
- Correct the light accumulation according to a ray-marched reference
- Avoid negative values of the sun disk illuminance when the sun disk is below the horizon

## Solution

- Adding a Newton's method iteration to `fast_sqrt` function
- Switched to using `fast_acos_4` for better precision of the sun angle towards the horizon (view mu angle = 0)
- Simplified the function for mapping to and from the Sky View UV coordinates by removing an if statement and correctly apply the method proposed by the [Hillarie paper](https://sebh.github.io/publications/egsr2020.pdf) detailed in section 5.3 and 5.4.
- Replaced the `ray_dir_ws.y` term with a shadow factor in the `sample_sun_illuminance` function that correctly approximates the sun disk occluded by the earth from any view point

## Testing

- Ran the atmosphere and SSAO examples to make sure the shaders still compile and run as expected.

---

## Showcase

<img width="1151" alt="showcase-img" src="https://github.com/user-attachments/assets/de875533-42bd-41f9-9fd0-d7cc57d6e51c" />

alice-i-cecile commented on crates/bevy_pbr/src/atmosphere/functions.wgsl:
```diff
@@ -71,41 +71,34 @@ fn sky_view_lut_r_mu_azimuth_to_uv(r: f32, mu: f32, azimuth: f32) -> vec2<f32> {
-    }
+    let view_zenith = fast_acos_4(mu);
+
+    // Latitude parameterization
```
This comment could be more clear for those who aren't familiar with this area.

alice-i-cecile commented on crates/bevy_pbr/src/atmosphere/functions.wgsl:
```diff
@@ -236,15 +243,18 @@ fn sample_local_inscattering(local_atmosphere: AtmosphereSample, ray_dir: vec3<f
         let light = &lights.directional_lights[light_i];
         let neg_LdotV = dot((*light).direction_to_light, ray_dir_ws);
-        let angle_to_sun = fast_acos(neg_LdotV);
+        let angle_to_sun = fast_acos_4(neg_LdotV);
```
A comment here about why acos_4 was chosen would be nice.

jnhyatt commented on crates/bevy_pbr/src/atmosphere/functions.wgsl:
```diff
@@ -127,12 +120,26 @@ fn sample_sky_view_lut(r: f32, ray_dir_as: vec3<f32>) -> vec3<f32> {
+    let view_ray_dist = length(view_pos.xyz / view_pos.w) * settings.scene_units_to_m;
+    let t_max = settings.aerial_view_lut_max_distance;
+
+    // Special handling for first slice to avoid extrapolation
```
I would clarify here just a little bit. Your explanation on Discord helped a lot

jnhyatt reviewed (approved):
Just a bit of clarification on top of alice's, otherwise lgtm. Did some additional testing, I'll post a video showing the from-space view of the sun disc correctly dropping below the planet edge rather than the 0 degree zenith.

jnhyatt commented:
https://github.com/user-attachments/assets/3c09a456-7a3c-4091-a14e-54335c33f409

ecoskey commented on crates/bevy_pbr/src/atmosphere/functions.wgsl:
```diff
@@ -127,12 +120,26 @@ fn sample_sky_view_lut(r: f32, ray_dir_as: vec3<f32>) -> vec3<f32> {
+        let f = view_ray_dist / delta_slice;
+        let first_slice_uvw = vec3(uv, 0.5 / f32(settings.aerial_view_lut_size.z));
+        let sample = textureSampleLevel(aerial_view_lut, aerial_view_lut_sampler, first_slice_uvw, 0.0);
+        return vec4(sample.rgb * f, pow(sample.a, f));
```
why the `pow` here?

ecoskey commented on crates/bevy_pbr/src/atmosphere/functions.wgsl:
```diff
@@ -127,12 +120,26 @@ fn sample_sky_view_lut(r: f32, ray_dir_as: vec3<f32>) -> vec3<f32> {
+    // Offset by 0.5 slice to sample at the center of each slice
+    let normalized_depth = (view_ray_dist / t_max) * f32(settings.aerial_view_lut_size.z - 1u);
+    let w = (normalized_depth - 0.5) / f32(settings.aerial_view_lut_size.z);
+    let uvw = vec3(uv, clamp(w, 0.0, 1.0));
```
maybe I'm missing something but this feels overcomplicated compared to just offsetting w by `0.5 / size.z`

ecoskey replied:
actually why do this at all? we're not doing any math with texel coords so it's not like we're trying to avoid sampling in between texels--that's the point right?

mate-h replied:
Good points here, I was looking into simplifying this function some more and avoiding that offset by 0.5 to get a good match to the ray marched reference, will continue playing around and add a commit for a simplification and fix.

mate-h replied:
After some testing and deliberation, I still settled on sampling the center of each LUT to produce matching result with the raymarched reference, this is because of better numerical precision when approximating the scattering integral using analytical approximation using the midpoint rule. Added code comment about this

ecoskey replied:
If it better matches the reference, I'm all for it. I'm a little confused though, because it seems like this would mean sampling potentially far away from where we're meant to be sampling. For example, with 32 1km slices and a sample 1km away from the camera, we'd ideally interpolate evenly between slices 0 and 1, which are sampled at 500m and 1500m respectively. But with this change we'd sample purely from the 500m slice, and only get about half the optical density, unless I'm misunderstanding 🤔

ecoskey reviewed (approved):
few small nits that could use clarification, otherwise lgtm :)

mate-h commented on crates/bevy_pbr/src/atmosphere/functions.wgsl:
```diff
@@ -127,12 +120,26 @@ fn sample_sky_view_lut(r: f32, ray_dir_as: vec3<f32>) -> vec3<f32> {
+        let f = view_ray_dist / delta_slice;
+        let first_slice_uvw = vec3(uv, 0.5 / f32(settings.aerial_view_lut_size.z));
+        let sample = textureSampleLevel(aerial_view_lut, aerial_view_lut_sampler, first_slice_uvw, 0.0);
+        return vec4(sample.rgb * f, pow(sample.a, f));
```
Because the transmittance needs to be interpolated according to Beer-Lambert law, I just realized that sampling the LUT with the linear sampler is incorrect at the end of this function too. So instead, I think we should store the `log(transmittance)` in the Aerial View LUT and then later recover it with `exp(sample)`

ecoskey replied:
makes sense 👍 

storing optical depth seems fine as long as it doesn't exceed the texel range, but that's probably pretty unlikely

mate-h commented:
![image](https://github.com/user-attachments/assets/643a1210-f996-41be-a45a-4061b3ef30dc)

mate-h commented:
After some more discussions, this discovery points to an issue in how the sample points are calculated. Therefore in order to remove the 0.5 texel offset in the sampling function, we need to fix the math that compute the sample positions. I will work on that tomorrow.

mate-h commented:
I've corrected the quadratic sampling in the sky view lut, reverted that to linear to recover the numerical accuracy. I've added a constant variable for the `MIDPOINT_RATIO` (open to renaming it) but it's now being used consistently in all raymarch loops. 
Additionally, I am using the `log(value)` and `exp(sample)` method to linearly sample the inscattered light in log space, so that gets rid of some of the banding at high altitudes for a lut that extends far.

mate-h commented on crates/bevy_pbr/src/atmosphere/aerial_view_lut.wgsl:
```diff
@@ -32,7 +32,10 @@ fn main(@builtin(global_invocation_id) idx: vec3<u32>) {
         for (var step_i: u32 = 0; step_i < settings.aerial_view_lut_samples; step_i++) {
-            let t_i = t_max * (f32(slice_i) + ((f32(step_i) + 0.5) / f32(settings.aerial_view_lut_samples))) / f32(settings.aerial_view_lut_size.z);
+            // Offset by -0.5 to align sampling position with slice boundaries, 
+            // since each texel stores the integral over its entire slice
```
I added this offset here and comment based on the discussion on discord, I think I am happy with this version overall.

mate-h replied:
![image](https://github.com/user-attachments/assets/fd2c85ac-a252-4ae8-9d2e-652e69ce718c)
for a bit more of a visual explanation as well, in this photo the red represents the center of each slice while the green is the boundaries in the texture coordinate space.

mate-h commented on crates/bevy_pbr/src/atmosphere/aerial_view_lut.wgsl:
```diff
@@ -55,8 +58,12 @@ fn main(@builtin(global_invocation_id) idx: vec3<u32>) {
-        textureStore(aerial_view_lut_out, vec3(vec2<u32>(idx.xy), slice_i), vec4(total_inscattering, mean_transmittance));
+
+        // Store the optical depth so that it can be linearly sampled between slices
+        let optical_depth = -log(max(mean_transmittance, 1e-6)); // Avoid log(0)
```
Could call this variable neg_optical_depth or log_transmittance and remove the minus from both here and the sample function, might make it more readable.

ecoskey replied:
agreed, that's definitely simpler 👍

mate-h commented on crates/bevy_pbr/src/atmosphere/functions.wgsl:
```diff
@@ -38,11 +38,17 @@
+// where in the segment that sample is taken (0.0 = start, 0.5 = middle, 1.0 = end).
+// We use 0.3 to sample closer to the start of each segment, which better approximates
+// the exponential falloff of atmospheric density.
+const MIDPOINT_RATIO: f32 = 0.3;
```
This could be exposed as a uniform variable, since it can differ depending on the scale height of the scatterers in the atmosphere.

ecoskey replied:
possibly, if it doesn't add more complexity to the user-facing api. Alternatively we could also just do the integration differently, for example piecewise linearly or quadratically (trapezoid rule or simpson's rule)

ecoskey commented on crates/bevy_pbr/src/atmosphere/aerial_view_lut.wgsl:
```diff
@@ -32,7 +32,10 @@ fn main(@builtin(global_invocation_id) idx: vec3<u32>) {
-            let t_i = t_max * (f32(slice_i) + ((f32(step_i) + 0.5) / f32(settings.aerial_view_lut_samples))) / f32(settings.aerial_view_lut_size.z);
+            // Offset by -0.5 to align sampling position with slice boundaries, 
+            // since each texel stores the integral over its entire slice
+            var t_i = t_max * (f32(slice_i) - 0.5 + ((f32(step_i) + MIDPOINT_RATIO) / f32(settings.aerial_view_lut_samples))) / f32(settings.aerial_view_lut_size.z);
```
I feel really bad for this back-and-forth, but I actually feel like doing the offset in `sample_aerial_view_lut` makes more sense, since then the texel values mean the same thing for each slice. It does remove the need for the special first-slice handling, but I feel like handling it explicitly is better than hiding it in how the lut is parametrized.

mate-h replied:
Okay let's be more explicit about that, agreed. I also updated a code comment to be more clear about the intent of the texel offset.

ecoskey commented on crates/bevy_pbr/src/atmosphere/functions.wgsl:
```diff
@@ -71,41 +77,38 @@ fn sky_view_lut_r_mu_azimuth_to_uv(r: f32, mu: f32, azimuth: f32) -> vec2<f32> {
+    // Apply non-linear transformation to compress more texels 
+    // near the horizon where high-frequency details matter most
+    // l is latitude in [-π/2, π/2] and v is texture coordinate in [0,1]
+    let l = view_zenith - horizon_zenith;
```
just wanted to mention how happy I am that we get to remove that mess inside the if-else from before :)

ecoskey commented on crates/bevy_pbr/src/atmosphere/sky_view_lut.wgsl:
```diff
@@ -36,19 +37,13 @@ fn main(@builtin(global_invocation_id) idx: vec3<u32>) {
 
     // Raymarch with quadratic distribution
```
```suggestion
```
Should remove this if we're back to midpoint rule

mate-h commented:
Okay great! I think I've addressed everything and this PR is ready for the merge queue @alice-i-cecile

ecoskey commented on crates/bevy_pbr/src/atmosphere/functions.wgsl:
```diff
@@ -135,10 +135,18 @@ fn sample_sky_view_lut(r: f32, ray_dir_as: vec3<f32>) -> vec3<f32> {
+    let fade = saturate(dist / delta_slice);
     // Recover the values from log space
-    return vec4(exp(sample.rgb), exp(-sample.a));
+    return vec4(exp(sample.rgb) * fade, exp(sample.a) * fade);
```
one tiny nit: could use `return exp(sample) * fade;` here :)

mate-h replied:
Good suggestion! Added ⚡️

alice-i-cecile commented:
Great cleanup and bug fixes; thanks a ton <3
